src/java.base/share/classes/sun/security/tools/keytool/Main.java

```diff
@@ -4748,6 +4890,61 @@ private boolean isTrustedCert(Certificate cert) throws KeyStoreException {
         }
     }
 
+    private void checkWeakConstraint(String label, String sigAlg, Key key,
+            CertPathConstraintsParameters cpcp) {
+        if (sigAlg != null) {
+            try {
+                DISABLED_CHECK.permits(sigAlg, cpcp, false);
+                try {
+                    LEGACY_CHECK.permits(sigAlg, cpcp, false);
+                } catch (CertPathValidatorException e) {
+                    weakWarnings.add(String.format(
+                            rb.getString("whose.sigalg.weak"), label, sigAlg));
+                }
+            } catch (CertPathValidatorException e) {
+                String eMessage = e.getMessage();
+                if (eMessage.contains("denyAfter constraint check failed") &&
+                        e.getReason() == BasicReason.ALGORITHM_CONSTRAINED) {
+                    String separator = "java.security: ";
```

Comment:
Fixed.

Reply:
After further thought, I'm now not sure my suggestion is any better (sorry for the rework). It is possible that there could be more than one `denyAfter` constraint, and in that case, you would need to also match on the algorithm that the constraint applies to, and that gets pretty complicated.

So, I now think your previous fix is probably better, even though it means we are depending on the syntax of the exception message. To avoid that from causing issues in the future, I would enhance your regression test to fail if the exception message changes in the future such that the denyAfter date cannot be parsed and is not what is expected.